Adds dark theme

shobhitagarwal1612 opened the pull request:
Closes #93 

#### What has been done to verify that this works as intended?

#### Why is this the best possible solution? Were any other approaches considered?

#### Are there any risks to merging this code? If so, what are they?

#### Do we need any specific form for testing your changes? If so, please attach one.

shobhitagarwal1612 committed:
Adds dark theme in user interface settings

shobhitagarwal1612 committed:
Updates activity's theme using shared prefs

shobhitagarwal1612 committed:
Adds dark theme for bottom dialog
Fixes text color for dark theme

shobhitagarwal1612 committed:
Adds ThemeUtils to manage themes for different items
Fixes dark theme for settings

shobhitagarwal1612 committed:
Fixes the tint color of menu icons

shobhitagarwal1612 commented:
@lognaturel @yanokwa and everyone else, this is still work in progress but it would be good to get a feedback to assure that I am going in the right direction.

![image](https://user-images.githubusercontent.com/8918023/37869513-ccc63b98-2fde-11e8-9983-1d0fffeeed5e.png)

![image](https://user-images.githubusercontent.com/8918023/37869520-d6dbc54e-2fde-11e8-8999-2f83f5f60ff0.png)

![image](https://user-images.githubusercontent.com/8918023/37869524-e55858d0-2fde-11e8-80db-a1467e90ab05.png)

![image](https://user-images.githubusercontent.com/8918023/37869532-f1533006-2fde-11e8-85ff-edcd7ecce293.png)

![image](https://user-images.githubusercontent.com/8918023/37869534-fb32ff48-2fde-11e8-8ef2-7ca2a9cce2e7.png)

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java:
```diff
@@ -111,112 +110,81 @@ protected void onCreate(Bundle savedInstanceState) {
+            if (Collect.allowClick()) {
+                Collect.getInstance().getActivityLogger()
+                        .logAction(this, "fillBlankForm", "click");
+                startActivity(new Intent(this, FormChooserList.class));
```
I love to see things getting shorter.

shobhitagarwal1612 committed:
Moves code to select theme for activities and menu to BaseActivity

shobhitagarwal1612 committed:
Fixes HierarchyView, AlertDialogs and status text

shobhitagarwal1612 commented:
@lognaturel @yanokwa This is ready for you to review!

shobhitagarwal1612 committed:
Adds themes to geo based activities

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/BaseActivity.java:
```diff
@@ -0,0 +1,40 @@
+
+import org.odk.collect.android.utilities.ThemeUtils;
+
+public abstract class BaseActivity extends AppCompatActivity {
```
We already have a base activity it's called CollectAbstractActivity so you can use it.

shobhitagarwal1612 replied:
Thanks for the suggestion. Things look much better now :smiley:

shobhitagarwal1612 committed:
Uses CollectAbstractActivity instead of BaseActivity

lognaturel commented:
The screenshots look great! I can't wait to see how/where users will put this to use.

Currently there's a crash with 
```
java.lang.ClassCastException: org.odk.collect.android.preferences.PreferencesActivity cannot be cast to org.odk.collect.android.activities.CollectAbstractActivity
	at org.odk.collect.android.preferences.UserInterfacePreferences.lambda$initThemePrefs$0$UserInterfacePreferences(UserInterfacePreferences.java:91)
	at org.odk.collect.android.preferences.UserInterfacePreferences$$Lambda$0.onPreferenceChange(Unknown Source)
	at android.preference.Preference.callChangeListener(Preference.java:1108)
	at android.preference.ListPreference.onDialogClosed(ListPreference.java:281)
	at android.preference.DialogPreference.onDismiss(DialogPreference.java:420)
	at android.app.Dialog$ListenersHandler.handleMessage(Dialog.java:1454)
```

shobhitagarwal1612 commented:
Did you try it out on your phone as well? The crash should be fixed now

lognaturel commented:
It looks great! 🎉 

I'm noticing the dialogs for the `no-calendar` date/time widgets are still white and in the spinner select one widget text is black on black but other than that I think you got it all. All your previous work getting things to use the latest APIs has really paid off.

It should be something that can be hidden from Admin Settings > User Settings.

I'm going to do a careful review of the commits that make unrelated changes to be sure we don't introduce regressions. I think we can get this in the next release -- that means it will be in while we do the post-freeze verification to make sure no regressions were introduced. The dark theme doesn't need to be tested very thoroughly now -- we can tell users that it's not fully verified and that they should report any issues that they find.

shobhitagarwal1612 commented:
> I'm noticing the dialogs for the no-calendar date/time widgets are still white and in the spinner select one widget text is black on black 

I totally forgot about this. Will fix this soon

shobhitagarwal1612 commented:
> It should be something that can be hidden from Admin Settings > User Settings.

+1

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -1617,27 +1611,7 @@ private void createConstraintToast(FormIndex index, int saveStatus) {
-    /**
-     * Creates a toast with the specified message.
-     */
-    private void showCustomToast(String message, int duration) {
```
This method displayed a toast in the center of a screen. I didn't like it but you definitely shouldn't change it as a part of this pr. I understand you because I've always struggled with not implementing many not related changes too :smile: but could you leave only those needed? I'm' talking about Lambda Expressions as well.

grzesiek2010 replied:
If we want lambda expression you can create a quick win issue since probably we can just replace all of them using Android Studio.

shobhitagarwal1612 replied:
The idea of removing this was to create two customized toast message theme which is controlled from within `ToastUtils`. Right now, we get a dark background with white text in both of the themes. Should a create a light/dark theme for this as well?

lognaturel replied:
Heh, it's so tricky to know when to do this kind of cleanup! I agree with you @grzesiek2010 that it's really important refactors be kept in their own commits. I think in this case they mostly are so we can review commit by commit and maintain history when merging. Unrelated changes are problematic when in the same commit or if a PR needs to be squashed.

In this specific case, you're right that the behavior changes slightly but I don't think the centering was meaningful. What do you think?

shobhitagarwal1612 replied:
Because in the dark theme, the screen is already dark and the Toast window is not as visible as it is in Light theme

shobhitagarwal1612 replied:
```
    <!-- Default Gravity setting for the system Toast view. Equivalent to: Gravity.CENTER_HORIZONTAL | Gravity.BOTTOM -->
    <integer name="config_toastDefaultGravity">0x00000051</integer>
```

This is the default gravity of the Toast

shobhitagarwal1612 replied:
Also, the textview's gravity is center_horizontal too by default

`transient_notification.xml`

shobhitagarwal1612 replied:
@lognaturel @grzesiek2010 So should I remove all the unrelated changes? (lamdas, code indentation, cleanup)

grzesiek2010 replied:
especially lambdas I think because as I said if we want to use lambdas let's fix all classes.

lognaturel replied:
Ok, thanks for the details, @shobhitagarwal1612, that helps make it clear this is not an unrelated change. I saw you used a detailed message in some commits, this kind of information about why you're doing something can be really helpful to put there (e.g. "All toasts go through ToastUtils so theme can be controlled in one place.")

As far as whether it does make sense to theme the toasts, I think they're legible enough as is. I have not seen white toasts anywhere.

grzesiek2010 replied:
When it comes to the background I didn't test your changes but for example here https://www.google.pl/search?q=android+dark+theme+toast&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiPgZnh84raAhVCLFAKHV6RD_AQ_AUICigB&biw=1855&bih=917#imgrc=6ITl-8K6iwRFwM:

dark background and dark toast is well visible for me.

lognaturel replied:
> So should I remove all the unrelated changes? (lamdas, code indentation, cleanup)

In this specific case since the commits are clean and the behavior easy to verify, I lean more towards leaving it in. But @grzesiek2010 is right that it makes it more time-consuming to verify the PR and riskier in general. If he feels strongly about it, best to make the diff cleaner.

grzesiek2010 replied:
It looks as if there is a white toast in Android 8 https://www.androidpolice.com/2017/10/26/android-8-1-feature-spotlight-toast-messages-now-white-better-match-rest-ui/

shobhitagarwal1612 replied:
Please correct me if I understood this wrong. So the plan is to revert back the lambdas but keep the Toast change, right? @lognaturel @grzesiek2010

shobhitagarwal1612 replied:
> It looks as if there is a white toast in Android 8 https://www.androidpolice.com/2017/10/26/android-8-1-feature-spotlight-toast-messages-now-white-better-match-rest-ui/

So this means the Toasts look different on different phones. Should we try to make it consistent for all or leave them as they are?

grzesiek2010 replied:
As @lognaturel said commits are quite clear so leave it if you want, just remember for the future. As I said I know The Boy Scout Rule and I also try to improve the code but in this pr, you have some completely unrelated like in HierarchyElement that's why I pointed this out.

grzesiek2010 replied:
> So this means the Toasts look different on different phones. Should we try to make it consistent for all or leave them as they are?

It's a default android behavior so we shouldn't change it.

shobhitagarwal1612 replied:
Great to know that there's a name for the rule too ` The Boy Scout Rule`
As mentioned @lognaturel,  @grzesiek2010 if the unrelated changes are looking fine to you as well then let's keep them as they are. I totally agree I did some completely unrelated changes which I shouldn't have and thanks for pointing it out to me. I need to be more careful while doing the refactors

shobhitagarwal1612 committed:
Moves duplicate code to start MainMenuActivity and close others

shobhitagarwal1612 committed:
Makes audio player compatible with dark theme

shobhitagarwal1612 committed:
Fixes date, time, ex widgets

shobhitagarwal1612 committed:
Adds option to hide setting in Admin > User Settings

shobhitagarwal1612 committed:
Fixes spinner widget text color

shobhitagarwal1612 commented:
@lognaturel Please review, I fixed the mentioned issues

shobhitagarwal1612 commented:
@lognaturel @yanokwa The geo-based activities are failing in master branch due to this [commit](https://github.com/opendatakit/collect/commit/a69adea105d23fd76206e0977d83e0db2879146f). I fixed that in the last commit

shobhitagarwal1612 committed:
Using ?colorControlNormal to control color of vector icons instead of
changing the tint using code.

shobhitagarwal1612 committed:
Revert lambdas changes in UserInterfacePreferences

shobhitagarwal1612 committed:
Revert unrealted changes in SortDialogAdapter

shobhitagarwal1612 commented:
@lognaturel @grzesiek2010 I have reverted back all the unrelated changes to make the PR easier to review and fixed the remaining issues as well

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java:
```diff
@@ -98,6 +99,7 @@
         // server_preferences.xml
         hashMap.put(KEY_PROTOCOL,                   Collect.getInstance().getString(R.string.protocol_odk_default));
         // user_interface_preferences.xml
+        hashMap.put(KEY_APP_THEME, Collect.getInstance().getString(R.string.app_theme_light));
```
Would be good to space the default over so it lines up nicely in a column with the others!

lognaturel reviewed (commented):
This looks very close to ready for QA to me! The one thing I've noticed is that `colorControlNormal` should be black and there should also be a tint for disabled which is currently the normal color. I have to say the black does look very stark but I think it's the right thing to do. CC @grzesiek2010 because I think he had thoughts on this as well.

The one exception is that I do like the gray-tinted icons for the exit dialog. If you'd rather make that black for consistency, I'm ok with that as well.

shobhitagarwal1612 commented:
> colorControlNormal should be black

I did try the black color too but the contrast between the icon color and the background looked too subtle and so went ahead with the default smoother colors. But these are just my gut feelings and we should do what's right. So I'd like others to weigh in as well.

lognaturel commented:
> the icon color and the background looked too subtle

Wondering if I wrote the wrong variable! I mean in the default light theme, things like the save or jump to icon on the form filling screen now look disabled. Black does look very high contrast but I think it's probably the right thing to do. Perhaps making the icons themselves more delicate would help but that's for another time.

shobhitagarwal1612 committed:
Sets the icon color as black/white for the light/dark theme

shobhitagarwal1612 committed:
Removes wrapper drawables

shobhitagarwal1612 committed:
Adds tests

shobhitagarwal1612 commented:
@lognaturel Please review again

shobhitagarwal1612 committed:
Resolves conflicts after rebasing

lognaturel commented:
The toasts have changed one last time so you will need to fix a merge conflict! Other than that, this is ready for some testing on different Android versions and devices. 👍

kkrawczyk123 commented:
Firstly I have to admit that this feature is fantastic! 

Tested with success on Androids 6.0, 7.0, 8.1. All screens display properly I tried all main functionalities and widgets. Import/Export Settings worked very well too.

Unfortunately, I wasn't able to run this build on Androids 4.1, 4.2 and 4.4 - Collect crashed. 

@opendatakit-bot unlabel "needs testing"

shobhitagarwal1612 commented:
Thanks for your thorough checks @kkrawczyk123 
I will try to reproduce the crash on an emulator

lognaturel commented:
Glad you like it, @kkrawczyk123, it is a super cool feature!

@shobhitagarwal1612 this is what I get:
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.odk.collect.android/org.odk.collect.android.activities.SplashScreenActivity}: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
at org.odk.collect.android.activities.SplashScreenActivity.onCreate(SplashScreenActivity.java:74)
```
I think the issue is that the `AndroidManifest` sets the theme.

shobhitagarwal1612 committed:
Fix crash on pre-lollipop devices

shobhitagarwal1612 commented:
@kkrawczyk123 @lognaturel Please review. I was able to fix the crash

shobhitagarwal1612 committed:
Reverts back the theme for Shortcut Activity

mmarciniak90 commented:
@opendatakit-bot label "needs testing"

shobhitagarwal1612 committed:
Replaces black icons with tint colored icons for Preference Screens

This was a fallback from replacing the icons with vectors. Vectors are working fine with lollipop and above devices but nothing is appearing in place of icons in pre-lollipop devices. I tried to create a custom view to enable the use of vectors but I went too far from the implementation of PreferencesActivity and it didn't feel like a good solution. So, until we have a better solution to replace the png's in preference-headers, we can continue to use the existing png's as drawables
''

mmarciniak90 commented:
@shobhitagarwal1612 great job! :thumbsup: This dark theme is awesome!! :smiley:

Unfortunately, I noticed that inactive buttons look almost the same as active. It is hard to distinguish them.

| new changes, dark theme | new changes, light theme  | master, without changes |
|---|---|---|
| ![screenshot_20180424-162006](https://user-images.githubusercontent.com/3907554/39189959-19ae3130-47c3-11e8-8bea-09c79fcf4a0f.png) | ![screenshot_20180424-161945](https://user-images.githubusercontent.com/3907554/39189989-2e0aa3f2-47c3-11e8-8305-c6e2471ad8b1.png) | ![screenshot_20180424-161736](https://user-images.githubusercontent.com/3907554/39190004-34a22afa-47c3-11e8-8f7c-727c46c24f5c.png)| 
| ![screenshot_20180424-162029](https://user-images.githubusercontent.com/3907554/39190030-41ff4afc-47c3-11e8-8f17-abb336ea330b.png) | ![screenshot_20180424-162044](https://user-images.githubusercontent.com/3907554/39190044-480d6aa0-47c3-11e8-9305-2c2bbec20cf8.png) | ![screenshot_20180424-162305](https://user-images.githubusercontent.com/3907554/39190056-4dc22ec2-47c3-11e8-996d-f4f0c16530a3.png)|

Also, texts in grey are in black now. @lognaturel do we want to keep this changes or restore to grey?

| new changes, light theme | master, without changes |
|---|---|
|![screenshot_20180424-161937](https://user-images.githubusercontent.com/3907554/39190168-8b0c7daa-47c3-11e8-8949-64ea89d26cca.png) | ![screenshot_20180424-161730](https://user-images.githubusercontent.com/3907554/39190170-8c480b3a-47c3-11e8-96b7-45b2bbbfb674.png)|

`Choose an account` dialog is light:

| | |
|---|---|
|![screenshot_20180424-164806](https://user-images.githubusercontent.com/3907554/39191720-18ded3a0-47c7-11e8-9bc3-873c14208e19.png)|![screenshot_20180424-165326](https://user-images.githubusercontent.com/3907554/39191724-1c60954a-47c7-11e8-8252-735ba132a135.png)|

`Save` icon is black in `Import/Export Settings` view
![screenshot_20180424-165054](https://user-images.githubusercontent.com/3907554/39191785-3f127b8a-47c7-11e8-98de-c7adf5b2b3a2.png)

yanokwa commented:
Thanks for catching these, @mmarciniak90! I do think we should restore the changes to grey.

shobhitagarwal1612 committed:
Colors only required text as black/white instead of all texts
 - Resolves button not looking as inactive
 - Restores gray secondary text

shobhitagarwal1612 committed:
Fixes theme color of account picker dialog

shobhitagarwal1612 committed:
Fixes save icon in import/export settings

shobhitagarwal1612 commented:
Verified on _Android 4.4, 7.1 and 8.0_

- [x] Restores text with gray color

- [x] Disabled buttons now appearing as disabled

- [x] Applies the correct theme to google account picker dialog 

- [x]  Save icon color fixed

@mmarciniak90 @yanokwa @lognaturel Please review

shobhitagarwal1612 committed:
Refactors ThemeUtils to make the code more testable

shobhitagarwal1612 committed:
Removes redundant app label

mmarciniak90 commented:
**Tested with success.**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

lognaturel commented:
👏 Awesome, @shobhitagarwal1612, this is a big one! Thank you @mmarciniak90 and @kkrawczyk123 for the thorough checks.

jd-alexander commented:
Awesome stuff! @shobhitagarwal1612 :100:  I was integrating another feature that needed to set the color of components and luckily I saw the PR and opted to use `colorControlNormal` and `themeUtils.getAttributeValue(R.attr.primaryTextColor)`. Do you guys think we should put this in the contribution guidelines as a means of enforcing this so that it will be easier for new features to adhere to this theming style? Or is this something that reviewers will just look out for.

shobhitagarwal1612 commented:
I think it's better if we put it in the pull request template as a checklist

lognaturel commented:
Good point, @jd-alexander, @shobhitagarwal1612. @huangyz0918 has started a PR checklist at https://github.com/opendatakit/collect/pull/2138 so we can add it in there.
